packages/theme-check-common/src/checks/undefined-object/index.ts

```diff
@@ -0,0 +1,156 @@
+import {
+  LiquidHtmlNode,
+  LiquidTag,
+  LiquidTagAssign,
+  LiquidTagCapture,
+  LiquidTagFor,
+  LiquidTagTablerow,
+  LiquidVariableLookup,
+  NamedTags,
+  NodeTypes,
+  Position,
+} from '@shopify/liquid-html-parser';
+import { LiquidCheckDefinition, Severity, SourceCodeType } from '../../types';
+import { last } from '../../utils';
+
+type Scope = { start?: number; end?: number };
+
+export const UndefinedObject: LiquidCheckDefinition = {
+  meta: {
+    code: 'UndefinedObject',
+    name: 'Undefined Object',
+    docs: {
+      description: 'This check exists to identify references to undefined Liquid objects.',
+      recommended: true,
+      url: 'https://shopify.dev/docs/themes/tools/theme-check/checks/undefined-object',
+    },
+    type: SourceCodeType.LiquidHtml,
+    severity: Severity.WARNING,
+    schema: {},
+    targets: [],
+  },
+
+  create(context) {
+    /**
+     * At present, snippet assets are not supported due to the inability of this
+     * check to handle objects defined in other assets.
+     */
+    if (context.relativePath(context.file.absolutePath).startsWith('snippets/')) {
+      return {};
+    }
+
+    const variableScopes: Map<string, Scope[]> = new Map();
+    const variables: LiquidVariableLookup[] = [];
+
+    function indexVariableScope(variableName: string | null, scope: Scope) {
+      if (!variableName) return;
+
+      const indexedScope = variableScopes.get(variableName) ?? [];
+      const currentScope = [...indexedScope, scope];
+
+      variableScopes.set(variableName, currentScope);
+    }
+
+    return {
+      async LiquidTag(node) {
+        if (isLiquidTagAssign(node)) {
+          indexVariableScope(node.markup.name, {
+            start: node.blockStartPosition.end,
+            end: node.blockEndPosition?.start,
+          });
+        }
+
+        if (isLiquidTagCapture(node)) {
+          indexVariableScope(node.markup.name, {
+            start: node.blockEndPosition?.end,
+          });
+        }
+
+        if (isLiquidForTag(node) || isLiquidTableRowTag(node)) {
+          indexVariableScope(node.markup.variableName, {
+            start: node.blockStartPosition.end,
+            end: node.blockEndPosition?.start,
+          });
+        }
+      },
+
+      async VariableLookup(node, ancestors) {
+        const parent = last(ancestors);
+
+        if (isLiquidTag(parent) && isLiquidTagCapture(parent)) return;
+
+        variables.push(node);
+      },
+
+      async onCodePathEnd() {
+        if (context.themeDocset) {
+          const objects = await context.themeDocset.objects();
+          objects.forEach((obj) => variableScopes.set(obj.name, []));
```

Comment:
You're missing a bit of `access` logic in here. See TypeSystem's in the language server. (e.g. `image` is only a type and never a "globally accessible" variable)

Reply:
Excellent call :) thank you! I've applied the `access` logic here :)